teerex/sgx-verify/src/lib.rs

```diff
@@ -0,0 +1,729 @@
+use webpki::SignatureAlgorithm;
+use x509_cert::Certificate;
+
+pub mod collateral;
+mod ephemeral_key;
+mod netscape_comment;
+#[cfg(test)]
+mod tests;
+mod utils;
+
+const SGX_REPORT_DATA_SIZE: usize = 64;
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+#[repr(C)]
+pub struct SgxReportData {
+	d: [u8; SGX_REPORT_DATA_SIZE],
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+#[repr(C)]
+pub struct SGXAttributes {
+	flags: u64,
+	xfrm: u64,
+}
+
+#[derive(Decode, Clone, TypeInfo)]
+#[repr(C)]
+pub struct DcapQuote {
+	header: DcapQuoteHeader,
+	body: SgxReportBody,
+	signature_data_len: u32,
+	quote_signature_data: EcdsaQuoteSignature,
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+#[repr(C)]
+pub struct DcapQuoteHeader {
+	version: u16,
+	attestation_key_type: u16,
+	reserved: u32,
+	qe_svn: u16,
+	pce_svn: u16,
+	qe_vendor_id: [u8; 16],
+	user_data: [u8; 20],
+}
+
+const ATTESTATION_KEY_SIZE: usize = 64;
+const REPORT_SIGNATURE_SIZE: usize = 64;
+
+#[derive(Decode, Clone, TypeInfo)]
+#[repr(C)]
+pub struct EcdsaQuoteSignature {
+	isv_enclave_report_signature: [u8; REPORT_SIGNATURE_SIZE],
+	ecdsa_attestation_key: [u8; ATTESTATION_KEY_SIZE],
+	qe_report: SgxReportBody,
+	qe_report_signature: [u8; REPORT_SIGNATURE_SIZE],
+	qe_authentication_data: QeAuthenticationData,
+	qe_certification_data: QeCertificationData,
+}
+
+#[derive(Clone, TypeInfo)]
+#[repr(C)]
+pub struct QeAuthenticationData {
+	size: u16,
+	certification_data: Vec<u8>,
+}
+
+impl Decode for QeAuthenticationData {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let mut size_buf: [u8; 2] = [0; 2];
+		input.read(&mut size_buf)?;
+		let size = u16::from_le_bytes(size_buf);
+
+		let mut certification_data = vec![0; size.into()];
+		input.read(&mut certification_data)?;
+
+		Ok(Self { size, certification_data })
+	}
+}
+
+#[derive(Clone, TypeInfo)]
+#[repr(C)]
+pub struct QeCertificationData {
+	certification_data_type: u16,
+	size: u32,
+	certification_data: Vec<u8>,
+}
+
+impl Decode for QeCertificationData {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let mut certification_data_type_buf: [u8; 2] = [0; 2];
+		input.read(&mut certification_data_type_buf)?;
+		let certification_data_type = u16::from_le_bytes(certification_data_type_buf);
+
+		let mut size_buf: [u8; 4] = [0; 4];
+		input.read(&mut size_buf)?;
+		let size = u32::from_le_bytes(size_buf);
+		// This is an arbitrary limit to prevent out of memory issues. Intel does not specify a max value
+		if size > 65_000 {
+			return Result::Err(codec::Error::from("Certification data too long"))
+		}
+
+		// Safety: The try_into() can only fail due to overflow on a 16-bit system, but we anyway
+		// ensure the value is small enough above.
+		let mut certification_data = vec![0; size.try_into().unwrap()];
+		input.read(&mut certification_data)?;
+
+		Ok(Self { certification_data_type, size, certification_data })
```

Comment:
Is there a way to "plug" custom traits int the scale codec? Because this is only a single field in a big hierarchical structure. For the most part this matches with what `Encode`/`Decode` does but unfortunately not in these two cases.

Reply:
Ah, I see. No, this is actually not possible, I guess we leave it as it is then. Thanks for the explanation!